app/presenters/tree_builder_services.rb

```diff
@@ -21,16 +21,12 @@ def root_options
 
   # Get root nodes count/array for explorer tree
   def x_get_tree_roots(count_only, _options)
-    roots = Rbac.filtered(Service.roots)
-
-    # Preload the root service picture since it's called by TreeNodeBuilder#build
-    MiqPreloader.preload(roots, :picture)
-
-    count_only_or_objects(count_only, roots, "name")
-  end
-
-  def x_get_tree_service_kids(object, count_only)
-    objects = Rbac.filtered(object.direct_service_children.select(&:display).sort_by { |o| o.name.downcase })
-    count_only_or_objects(count_only, objects, 'name')
+    all_services = Rbac.filtered(Service.where(:display => true))
+    if count_only
+      all_services.size
+    else
+      MiqPreloader.preload(all_services.to_a, :picture)
+      Service.arrange_nodes(all_services.sort_by { |n| [n.ancestry.to_s, n.name.downcase] })
```

Comment:
For some reason I want to offload on ruby. Maybe because the core ancestry sorts are so inefficient. But when running numbers in this case they were comparable (we do not use ancestry default sorting)